src/main/java/net/neoforged/neoforge/transfer/handlers/resources/IResourceHandler.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.handlers.resources;
+
+import net.neoforged.neoforge.transfer.handlers.ITransactionHandler;
+import net.neoforged.neoforge.transfer.handlers.templates.resource.ResourceStorageHandler;
+import net.neoforged.neoforge.transfer.resources.IResource;
+import net.neoforged.neoforge.transfer.transaction.SnapshotJournal;
+import net.neoforged.neoforge.transfer.transaction.TransactionContext;
+
+/**
+ * A generic handler for handling a {@link IResource resource} of type {@link T} whether it be inserting, extracting, querying some size value, etc.
+ * Trying to interact with a handler with an index larger than the size is expected to often times throw an {@link IndexOutOfBoundsException} or similar error. Ensure you are within the size constraints of the handler before calling a method.
+ *
+ * @param <T> The type of resource this handler manages.
+ */
+public interface IResourceHandler<T extends IResource> extends ITransactionHandler {
+    /**
+     * An index is synonymous with "slot", "tank", "buffer", etc.
+     *
+     * @return The number of indices this handler manages. <strong>Must be non-negative</strong>
+     */
+    int size();
+
+    /**
+     * @param index The index to get the resource from. <strong>Must be non-negative</strong>
+     * @return The resource at the given index.
+     */
+    T getResource(int index);
+
+    /**
+     * @param index The index to get the amount from. <strong>Must be non-negative</strong>
+     * @return The amount of the resource at the given index. <strong>Must be non-negative</strong>
+     * @see #getAmountAsLong(int)
+     */
+    int getAmount(int index);
+
+    /**
+     * @param index The index to get the amount from. <strong>Must be non-negative</strong>
+     * @return The amount as a long of the resource at the given index. <strong>Must be non-negative</strong>
+     * @see #getAmount(int) (int)
+     */
+    default long getAmountAsLong(int index) {
+        return getAmount(index);
+    }
+
+    /**
```

Comment:
We should specify the expected behavior for when the resource doesn't match the stored one (just return the value of the resource passed even if it isn't what is currently stored). We also should probably specify if we allow passing resources to this that return false from `isValid`